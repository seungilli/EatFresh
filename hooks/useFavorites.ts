import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Meal } from '../types/data';

const useFavorites = () => {
    const [favorites, setFavorites] = useState<Meal[]>([]);

    const fetchFavorites = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('favorite');
            const parsedFavorites = value != null ? JSON.parse(value) : [];
            setFavorites(Array.isArray(parsedFavorites) ? parsedFavorites : []);
        } catch (error) {
            console.error('Error fetching favorites:', error);
        }
    }, []);

    const addFavorite = useCallback(
        async (item: Meal) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (!Array.isArray(updatedFavorites)) {
                    updatedFavorites = [];
                }
                updatedFavorites.push(item);
                await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                await fetchFavorites();
            } catch (error) {
                console.error('Error adding favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const removeFavorite = useCallback(
        async (id: string) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (Array.isArray(updatedFavorites)) {
                    updatedFavorites = updatedFavorites.filter((meal: Meal) => meal.idMeal !== id);
                    await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                    await fetchFavorites();
                }
            } catch (error) {
                console.error('Error removing favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const clearFavorites = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('favorite');
            await fetchFavorites();
        } catch (error) {
            console.error('Error clearing favorites:', error);
        }
    }, [fetchFavorites]);

    useEffect(() => {
        fetchFavorites();
    }, [fetchFavorites]);

    return {
        favorites,
        addFavorite,
        removeFavorite,
        clearFavorites,
        refreshFavorites: fetchFavorites,
    };
};

export default useFavorites;
