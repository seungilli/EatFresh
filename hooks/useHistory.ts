import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { HistoryType } from '@/types/data';

const useHistory = () => {
    const [history, setHistory] = useState<HistoryType[]>([]);

    const fetchHistory = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('history');
            const parsedHistory = value != null ? JSON.parse(value) : [];
            setHistory(Array.isArray(parsedHistory) ? parsedHistory : []);
        } catch (error) {
            console.error('Error fetching history:', error);
        }
    }, []);

    const addHistoryItem = useCallback(
        async (item: HistoryType) => {
            try {
                const existingHistory = await AsyncStorage.getItem('history');
                let updatedHistory = existingHistory ? JSON.parse(existingHistory) : [];
                if (!Array.isArray(updatedHistory)) {
                    updatedHistory = [];
                }
                updatedHistory.push(item);
                await AsyncStorage.setItem('history', JSON.stringify(updatedHistory));
                await fetchHistory();
            } catch (error) {
                console.error('Error adding history:', error);
            }
        },
        [fetchHistory]
    );

    const clearHistory = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('history');
            await fetchHistory();
        } catch (error) {
            console.error('Error clearing history:', error);
        }
    }, [fetchHistory]);

    useEffect(() => {
        fetchHistory();
    }, [fetchHistory]);

    return {
        history,
        addHistoryItem,
        clearHistory,
        refreshHistory: fetchHistory,
    };
};

export default useHistory;
