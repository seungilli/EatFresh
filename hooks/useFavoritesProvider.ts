import React, { createContext, useContext, ReactNode } from "react";
import useFavorites from "./useFavorites";

type FavoritesContextType = ReturnType<typeof useFavorites>;

const FavoritesContext = createContext<FavoritesContextType | undefined>(
    undefined
);

export const FavoritesProvider = ({ children }: { children: ReactNode }) => {
    const favorites = useFavorites();

    return React.createElement(
        FavoritesContext.Provider,
        { value: favorites },
        children
    );
};

export const useFavoritesContext = (): FavoritesContextType => {
    const context = useContext(FavoritesContext);
    if (!context) {
        throw new Error(
            "useFavoritesContext must be used within a FavoritesProvider"
        );
    }
    return context;
};
