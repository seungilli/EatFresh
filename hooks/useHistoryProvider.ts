import React, { createContext, useContext, ReactNode } from "react";
import useHistory from "./useHistory";

type HistoryContextType = ReturnType<typeof useHistory>;

const HistoryContext = createContext<HistoryContextType | undefined>(
    undefined
);

export const HistoryProvider = ({ children }: { children: ReactNode }) => {
    const history = useHistory();

    return React.createElement(
        HistoryContext.Provider,
        { value: history },
        children
    );
};

export const useHistoryContext = (): HistoryContextType => {
    const context = useContext(HistoryContext);
    if (!context) {
        throw new Error(
            "useHistoryContext must be used within a HistoryProvider"
        );
    }
    return context;
};
