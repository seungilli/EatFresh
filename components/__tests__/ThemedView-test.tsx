import * as React from "react";
import renderer from "react-test-renderer";
import { ThemedView } from "../ThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";

jest.mock("@/hooks/useThemeColor", () => ({
  useThemeColor: jest.fn(),
}));

describe("<ThemedView />", () => {
  it("should render correctly with default props", () => {
    (useThemeColor as jest.Mock).mockReturnValue("white");

    const tree = renderer.create(<ThemedView />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should apply the correct background color from lightColor", () => {
    (useThemeColor as jest.Mock).mockReturnValue("lightblue");

    const tree = renderer
      .create(<ThemedView lightColor="lightblue" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should apply the correct background color from darkColor", () => {
    (useThemeColor as jest.Mock).mockReturnValue("darkblue");

    const tree = renderer.create(<ThemedView darkColor="darkblue" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should merge custom styles with themed background color", () => {
    (useThemeColor as jest.Mock).mockReturnValue("white");

    const tree = renderer
      .create(<ThemedView style={{ padding: 10, margin: 5 }} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
    expect(tree).toHaveProperty("props.style", [
      { backgroundColor: "white" },
      { padding: 10, margin: 5 },
    ]);
  });

  it("should pass other props to the View component", () => {
    (useThemeColor as jest.Mock).mockReturnValue("white");

    const testID = "themed-view";
    const tree = renderer.create(<ThemedView testID={testID} />).toJSON();

    expect(tree).toHaveProperty("props.testID", testID);
  });

  it("should handle undefined lightColor and darkColor gracefully", () => {
    (useThemeColor as jest.Mock).mockReturnValue(null);

    const tree = renderer.create(<ThemedView />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
