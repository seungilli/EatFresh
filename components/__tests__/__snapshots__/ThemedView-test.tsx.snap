// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#ffffff",
      },
      undefined,
    ]
  }
>
  Snapshot test!
</View>
`;
